reference/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,354 @@
+inline void update_s(
+    const gko::batch::multi_vector::batch_item<const ValueType>& r_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& alpha_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& v_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& s_entry)
+{
+    for (int r = 0; r < s_entry.num_rows; r++) {
+        s_entry.values[r * s_entry.stride] =
+            r_entry.values[r * r_entry.stride] -
+            alpha_entry.values[0] * v_entry.values[r * v_entry.stride];
+    }
+}
+
+
+template <typename ValueType>
+inline void compute_omega(
+    const gko::batch::multi_vector::batch_item<const ValueType>& t_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& s_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& temp_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& omega_entry)
+{
+    compute_dot_product_kernel<ValueType>(t_entry, s_entry, omega_entry);
+    compute_dot_product_kernel<ValueType>(t_entry, t_entry, temp_entry);
+    omega_entry.values[0] /= temp_entry.values[0];
+}
+
+
+template <typename ValueType>
+inline void update_x_and_r(
+    const gko::batch::multi_vector::batch_item<const ValueType>& p_hat_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& s_hat_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& alpha_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& omega_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& s_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& t_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& x_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& r_entry)
+{
+    const ValueType omega = omega_entry.values[0];
+    for (int r = 0; r < x_entry.num_rows; r++) {
+        x_entry.values[r * x_entry.stride] =
+            x_entry.values[r * x_entry.stride] +
+            alpha_entry.values[0] * p_hat_entry.values[r * p_hat_entry.stride] +
+            omega * s_hat_entry.values[r * s_hat_entry.stride];
+
+        r_entry.values[r * r_entry.stride] =
+            s_entry.values[r * s_entry.stride] -
+            omega * t_entry.values[r * t_entry.stride];
+    }
+}
+
+template <typename ValueType>
+inline void update_x_middle(
+    const gko::batch::multi_vector::batch_item<const ValueType>& alpha_entry,
+    const gko::batch::multi_vector::batch_item<const ValueType>& p_hat_entry,
+    const gko::batch::multi_vector::batch_item<ValueType>& x_entry)
+{
+    for (int r = 0; r < x_entry.num_rows; r++) {
+        x_entry.values[r * x_entry.stride] =
+            x_entry.values[r * x_entry.stride] +
+            alpha_entry.values[0] * p_hat_entry.values[r * p_hat_entry.stride];
+    }
+}
+
+
+template <typename StopType, typename PrecType, typename LogType,
+          typename BatchMatrixType, typename ValueType>
+inline void batch_entry_bicgstab_impl(
```

Comment:
I also think the core part can be shared among backends, but I do not focus on that now.
I assume the fused kernel from GPU perspective

Reply:
Yes, I think we can think about unifying this later.